zebra-state/src/service.rs

```diff
@@ -271,87 +289,142 @@ impl StateService {
+    /// chain. If there is no intersection with the best chain, starts from the genesis hash.
+    ///
+    /// Includes finalized and non-finalized blocks.
     ///
-    /// Only matches and returns hashes from the best chain, including non-finalized blocks.
     /// Stops the list of hashes after:
     ///   * adding the non-finalized best tip,
-    ///   * adding the stop hash to the list, if it is in the best chain, or
-    ///   * adding 500 hashes to the list.
-    pub fn find_chain_hashes(
+    ///   * adding the `stop` hash to the list, if it is in the best chain, or
+    ///   * adding `max_len` hashes to the list.
+    ///
+    /// Returns an empty list if the state is empty.
+    pub fn collect_chain_hashes(
         &self,
-        known_blocks: Vec<block::Hash>,
+        intersection: Option<block::Hash>,
         stop: Option<block::Hash>,
+        max_len: usize,
     ) -> Vec<block::Hash> {
-        let mut res: Vec<block::Hash> = Vec::new();
-        // Get chain tip:
-        // We can get a block locator request before we have downloaded the genesis block,
-        // if this is the case we return an empty response.
-        let chain_tip = self.tip();
-        if chain_tip.is_none() {
-            return res;
-        }
-        let chain_tip_height = chain_tip.unwrap().0;
-
-        // Get requested tip:
-        // Use the genesis hash if locator is empty or if no hash in the
-        // locator is present in our best chain.
-        let mut locator_tip_hash = genesis_hash(self.network);
-        let mut locator_tip_height = block::Height(1);
-        if !known_blocks.is_empty() {
-            for hash in known_blocks {
-                let height_hash = self.height_by_hash(hash);
-                if let Some(height_hash) = height_hash {
-                    locator_tip_hash = hash;
-                    locator_tip_height = height_hash;
-                    break;
-                }
-            }
-        }
-        tracing::info!(
-            "REQUESTED TIP: {:?} {:?}",
-            locator_tip_height,
-            locator_tip_hash
-        );
+        assert!(max_len > 0, "max_len must be at least 1");
 
-        const MAX_FIND_BLOCK_HASHES_RESULTS: usize = 500;
-
-        // Compute a new tip, make sure it is below our chain tip.
-        let new_tip_height = block::Height(std::cmp::min(
-            locator_tip_height.0 + MAX_FIND_BLOCK_HASHES_RESULTS as u32,
-            chain_tip_height.0,
-        ));
-        let new_tip_hash = self.hash(new_tip_height).expect("new tip must have a hash");
-
-        // Get a new chain starting at the new "requested" tip.
-        let new_chain = self.chain(new_tip_hash);
+        // We can get a block locator request before we have downloaded the genesis block
+        let chain_tip_height = self.tip().map(|t| t.0);
+        if chain_tip_height.is_none() {
+            return Vec::new();
+        }
+        let chain_tip_height = chain_tip_height.unwrap();
```

Comment:
NIT: I'd normally try to avoid an unwrap here structurally

```suggestion
        let chain_tip_height = if let Some((height, _)) = self.tip() {
            height
        } else {
            return Vec::new();
        };
```